### ISSUE TITLE
in-toto attestation JSON schema verify via CLI

### ISSUE THREAD

**Issue by developer-guy** [in-toto/in-toto-golang#129]
## Abstract
in-toto attestations are in the form of JSON. We can use [JSON schema](https://json-schema.org) to validate the schema of the attestation. 

## Motivation
We started to work on a [PR](https://github.com/sigstore/cosign/issues/512) to verify in-toto attestations by using Cue or Rego languages in the [cosign](https://github.com/sigstore/cosign) project. So, if in-toto attestations have JSON schemas, we can use them in the `in-toto-golang` CLI. Also, cosign has a dependency on the `in-toto-golang` project for the structs of the in-toto attestations. So, if we add support of verifying JSON schemas to the `in-toto-golang` project, we can use it in cosign project too.

## References

I found some additional resources that might help us to implement this:

* [gojsonschema](https://github.com/xeipuuv/gojsonschema)
* [JSON Schema Generators](https://json-schema.org/implementations.html#generators-from-schemas)
* [Go JSON Schema Reflection](https://github.com/alecthomas/jsonschema)

cc: @dentrax @erkanzileli

**Comment by colek42** [comment]
in-toto uses Canonical JSON which may complicate this effort.  http://wiki.laptop.org/go/Canonical_JSON

**Comment by shibumi** [comment]
This is a good point @colek42!

**Comment by developer-guy** [comment]
is it a blocker thing, or do we figure this out somehow? @shibumi @colek42

**Comment by shibumi** [comment]
Are Cue and JSON schemas canonical? If they are, I don't think this is a problem, right?

**Comment by developer-guy** [comment]
what do you mean by canonical? I really don't know how to respond that 🤷🏻‍♂️

**Comment by shibumi** [comment]
@developer-guy Sorry, my bad. Let me rephrase it: are CUE and JSON Schema compatible with Canonical JSON? If I get this right:

Cue and JSON Schema provide a schema for cue or JSON. Then I am able to validate the CUE/JSON against the Schema.
My question is: Can Cue or JSON Schema validate if the provided JSON for schema validation is Canonical JSON?

It is possible that my question is completely out of this discussion :D I am really not an expert in CUE or Canonical JSON.

**Comment by developer-guy** [comment]
sorry for the ping @verdverm but maybe you can help us here with @shibumi's question because I also don't know that. Btw, this topic is related to what we want to do in the [PR](https://github.com/sigstore/cosign/issues/512)  for the `cosign` project.

**Comment by verdverm** [comment]
From what I understand, Canonical JSON is a subset (restricted version) of JSON, though poorly named if you want my opinion ;]

Both CUE and JSON Schema should be able to handle this.

**Comment by shibumi** [comment]
@verdverm I understood Canonical JSON as "ordered" JSON. Question is if cue would validate this correctly:

file 1
```json
{
  "test": "A",
  "foo": 1;
}
```

file 2
```json
{
  "foo": 1;
  "test": "A";
}
```

Template JSON Schema or CUE:
```json
{
  "foo": number;
  "test": string;
}
```

With Canonical JSON only the file 2 should be approved by the template. The validation for file 1 MUST fail.
With normal JSON this is not the case and a JSON Schema would validate both files successfully, iirc.

**Comment by verdverm** [comment]
Ah, I missed that lexicographical part in their spec.

You could do this with CUE, though it would require some extra stuff. By default, CUE is intentionally order ignorant. You can probably use extra constraints (IsSorted on the object keys through a list comprehension)

**Comment by shibumi** [comment]
> Wouldn't generating attestations using DSSE help here? There's already some support for it...

I think the plan is to **validate** attestations via a schema, not to generate them. I might be wrong.

**Comment by verdverm** [comment]
FYI: https://github.com/cue-lang/cue/discussions/1285
